### ISSUE TITLE
[Feature]: vLLM DP=2 didn't speed up the training as low batch size.

### ISSUE THREAD

**Issue by glorgao** [vllm-project/vllm#17129]
### 🚀 The feature, motivation and pitch

Hi team,

First of all, thanks for the recent efforts, especially to @qgallouedec, for supporting the new `data_parallel_size` feature in vLLM. I tested the `vllm-serve-dp` branch with `data_parallel_size=2`, and confirmed that it launches two processes for rollouts as expected. Great work!

However, **the speedup of having `data_parallel_size=2` isn't quite as significant as I hoped**. In my previous setup using a single GPU, I was achieving around 4000–6000 toks/s generation. With `data_parallel_size=2`, this drops to only 1000–1200 tokens/s per process, which results in an overall slower or comparable throughput.

It seems that the GPUs may be underutilized, possibly waiting on input (prompts/questions) to arrive. I suspect the issue could be mitigated by **allowing larger batch sizes for generation in the vLLM server, while keeping a smaller batch size for gradient calculations to avoid OOM errors.**

I’m reporting this primarily for visibility, and I’m confident that the team can fine-tune the implementation to achieve near-linear speedup. Looking forward to seeing future improvements!

Thanks again for your hard work!

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by qgallouedec** [comment]
> thanks for the recent efforts, especially to @qgallouedec, for supporting the new data_parallel_size feature in vLLM

to be fair I'm not responsible for this 😂

In my experience, the speedup depends a lot on the model size, the max sequence length etc. Can you provide more information on your setup?

**Comment by glorgao** [comment]
**DP=1**

`CUDA_VISIBLE_DEVICES=0 python -m trl.scripts.vllm_serve --model Qwen/Qwen2.5-7B  --tensor_parallel_size 1 --data_parallel_size 1`

> Processed prompts: 100%|███████████| 336/336 [00:28<00:00, 11.84it/s, est. speed input: 2213.85 toks/s, output: 8083.83 toks/s]
INFO:     127.0.0.1:43830 - "POST /generate/ HTTP/1.1" 200 OK                                | 9/355 [04:30<3:09:34, 32.87s/it

Evaluating speed: 
**8083.83 toks/s, 32.87s/it**

```yaml
model_name_or_path: Qwen/Qwen2.5-7B
model_revision: main
torch_dtype: bfloat16
attn_implementation: flash_attention_2

bf16: true
beta: 0.0
epsilon_high: 0.28
use_vllm: true
do_eval: true
eval_strategy: 'steps'
eval_steps: 1
...
max_completion_length: 3072
...
num_generations: 14
per_device_eval_batch_size: 48
per_device_train_batch_size: 32
```

**Comment by glorgao** [comment]
**DP=2**

`CUDA_VISIBLE_DEVICES=0,1 python -m trl.scripts.vllm_serve --model Qwen/Qwen2.5-7B  --tensor_parallel_size 1 --data_parallel_size 2`

> Processed prompts: 100%|███████████| 144/144 [00:25<00:00,  5.71it/s, est. speed input: 1004.81 toks/s, output: 3497.78 toks/s]
  1%|█▎                                                                                      | 7/474 [03:05<3:49:46, 29.52s/it

Evaluating speed (I report the evaluation speed since for checking the speed difference in generating rollouts):
**3497.78 toks/s, 29.52s/it**
```yaml
model_name_or_path: Qwen/Qwen2.5-7B
model_revision: main
torch_dtype: bfloat16
attn_implementation: flash_attention_2

bf16: true
beta: 0.0
epsilon_high: 0.28
use_vllm: true
do_eval: true
eval_strategy: 'steps'
eval_steps: 1
...
max_completion_length: 3072
...
num_generations: 16
per_device_eval_batch_size: 48
per_device_train_batch_size: 32

```

**Comment by glorgao** [comment]
When `data_parallel_size=2`, the batch size (as shown in processed prompts from tqdm) is **144**, whereas with `data_parallel_size=1`, it's `336`. It seems that the rollout configuration becomes overly conservative when using multiple GPUs.

**Comment by glorgao** [comment]
Apologies! I mistakenly opened this issue under the vLLM project — it was actually meant for the `trl` repo. My bad! >_<
